sql: refactor privilege to remove eval dep from catalog

RichardJCai opened the pull request:
Refactor to break dependency + bugfix

Release note (bug fix): Previously, when running UNION on with a
show backup query, it could non-deterministically cause a panic due to
a race condition. This is now fixed.

Example:
SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'] UNION SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'];

Resolves https://github.com/cockroachdb/cockroach/issues/88385
Resolves https://github.com/cockroachdb/cockroach/issues/88400

RichardJCai committed:
sql: fix panic when running UNION on SHOW BACKUP query

This refactor fixes the panic and also removes the
eval dep from catalog

Release note (bug fix): Previously, when running UNION on with a
show backup query, it could non-deterministically cause a panic due to
a race condition. This is now fixed.

Example:
SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'] UNION SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'];

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/88787)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> Reviewed 46 of 53 files at r1, 1 of 1 files at r2, 2 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88787)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin and @RichardJCai)

___
*[`pkg/ccl/backupccl/show.go` line 293 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4ZkyX40qultePy4UF:-ND4ZkyX40qultePy4UG:b-mmhbth) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/ccl/backupccl/show.go#L293)):*
> ```Go
> 			return err
> 		}
> 		_, _, err = p.ResolveMutableTableDescriptor(ctx, externalconn.SystemExternalConnectionsTableName, true, tree.ResolveRequireTableDesc)
> ```

Isn't mutable here wrong? Should this be immutable? Mutable means avoiding leases. 



_Code quote:_
```Go
		// Resolve the system.privileges and system.external_connections
		// tables such that we have leases on those two tables such that
		// we don't have to resolve them during Privilege checks.
		// Checking them in the privilege check below causes a race
		// condition, the collection should not be accessed during
		// execution time.
		_, _, err = p.ResolveMutableTableDescriptor(ctx, syntheticprivilege.SystemPrivilegesTableName, true, tree.ResolveRequireTableDesc)
		if err != nil {
			return err
		}
		_, _, err = p.ResolveMutableTableDescriptor(ctx, externalconn.SystemExternalConnectionsTableName, true, tree.ResolveRequireTableDesc)
```
___
*[`pkg/sql/synthesize_privileges.go` line 163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4Vgfa16E5L6C5NmHI:-ND4Vgfa16E5L6C5NmHJ:bq2emsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fdf0ad2cdd6644c8463877ef6e36e4ddd79b603/pkg/sql/synthesize_privileges.go#L163)):*
> ```Go
> 			// We use InvalidID to skip checks on the root/admin roles having
> 			// privileges.
> 			if err := privileges.Validate(descpb.InvalidID, privilegeObjectType, privilegeObjectPath, privilege.GetValidPrivilegesForObject(privilegeObjectType)); err != nil {
> ```

nit: wrap this
___
*[`pkg/sql/synthesize_privileges.go` line 64 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_0qf23GG6S66neJT:-ND4_0qf23GG6S66neJU:b3dkcm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L64)):*
> ```Go
> 	var tableVersions []descpb.DescriptorVersion
> 	cache := p.syntheticPrivilegeCache
> 	found, privileges, retErr := func() (bool, *catpb.PrivilegeDescriptor, error) {
> ```

there's no error that ever gets returned here, right?
___
*[`pkg/sql/synthesize_privileges.go` line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_8_A5MGGUuqnQ4Df:-ND4_8_A5MGGUuqnQ4Dg:be9wpop) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L83)):*
> ```Go
> 	}
> 
> 	val, err := cache.LoadValueOutsideOfCache(ctx, privilegeObjectPath, func(loadCtx context.Context) (_ interface{}, retErr error) {
> ```

why name this error if you don't use it?
___
*[`pkg/sql/synthesize_privileges.go` line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_B9I4HoZ-Ke7nW1I:-ND4_B9I4HoZ-Ke7nW1J:b-o94mbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L83)):*
> ```Go
> 	}
> 
> 	val, err := cache.LoadValueOutsideOfCache(ctx, privilegeObjectPath, func(loadCtx context.Context) (_ interface{}, retErr error) {
> ```

nit wrap this at `func(`
___
*[`pkg/sql/synthesize_privileges.go` line 97 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_Qba0VDeB6LOCkuG:-ND4_Qba0VDeB6LOCkuH:bpzz9o6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L97)):*
> ```Go
> 			}
> 			defer func() {
> 				retErr = errors.CombineErrors(retErr, it.Close())
> ```

shouldn't this be doing this to the inner error and not the outer error?
___
*[`pkg/sql/synthesize_privileges.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_NzY1lTlbpuLaHnT:-ND4_NzY1lTlbpuLaHnU:bd98xzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L101)):*
> ```Go
> 
> 			privileges = &catpb.PrivilegeDescriptor{}
> 			for {
> ```

nit: consider pulling this for loop into a helper



_Code quote:_
```Go
			privileges = &catpb.PrivilegeDescriptor{}
			for {
```
___
*[`pkg/sql/catalog/descs/collection.go` line 141 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_hc486T0SyenGlSS:-ND4_hc486T0SyenGlST:b-6bm21a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/catalog/descs/collection.go#L141)):*
> ```Go
> 	sqlLivenessSession sqlliveness.Session
> 
> 	privilegeSynthesizer syntheticprivilege.PrivilegeSynthesizer
> ```

give this some commentary while you're here. What is this? What is it used for? etc. 
___
*[`pkg/sql/catalog/tabledesc/table_desc.go` line 650 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4aX3yDLPvEXKRCz5k:-ND4aX3yDLPvEXKRCz5l:b-cp7vv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/catalog/tabledesc/table_desc.go#L650)):*
> ```Go
> func (desc *wrapper) GetPrivileges() *catpb.PrivilegeDescriptor {
> 	if desc.IsVirtualTable() && desc.TableDesc().Privileges == nil {
> 		return catpb.NewPrivilegeDescriptor(
> ```

Can you put commentary around this case? When does it happen? Should it happen? 



_Code quote:_
```Go
			username.PublicRoleName(), privilege.List{privilege.USAGE},
			privilege.List{}, username.NodeUserName())
```
___
*[`pkg/sql/sem/eval/deps.go` line 345 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4aicA2Erwnqrqu_DG:-ND4aicA2Erwnqrqu_DH:bncstcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/sem/eval/deps.go#L345)):*
> ```Go
> 	// PrivilegeDescriptor given a SyntheticPrivilegeObject's path
> 	// from system.privileges.
> 	SynthesizePrivilegeDescriptor(
> ```

What does this get used for? 
___
*[`pkg/sql/syntheticprivilege/privilege_synthesizer.go` line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4aq6AEKy350FrM3oe:-ND4aq6AEKy350FrM3of:b-9wyec5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/syntheticprivilege/privilege_synthesizer.go#L28)):*
> ```Go
> 		ctx context.Context, name string, privilegeObjectType privilege.ObjectType, tableVersion descpb.DescriptorVersion,
> 	) (*catpb.PrivilegeDescriptor, error)
> 	SetIEFactory(ieFactory sqlutil.InternalExecutorFactory)
> ```

what is this about? 


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88787)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @benbardin)

___
*[`pkg/sql/synthesize_privileges.go` line 64 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_0qf23GG6S66neJT:-NDinRVLD5FJLSvRwQ1s:bu0rs8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L64)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

there's no error that ever gets returned here, right?
</blockquote></details>

Yep, removed
___
*[`pkg/sql/synthesize_privileges.go` line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_8_A5MGGUuqnQ4Df:-NDinSES6eDJM6W5gmi5:ba3dfob) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L83)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why name this error if you don't use it?
</blockquote></details>

Removed
___
*[`pkg/sql/synthesize_privileges.go` line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_B9I4HoZ-Ke7nW1I:-NDinfUh3WnHdMTKPDk4:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L83)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit wrap this at `func(`
</blockquote></details>

Done
___
*[`pkg/sql/synthesize_privileges.go` line 97 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_Qba0VDeB6LOCkuG:-NDino4g0kn7s_FkcDt6:b-vxv7gt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L97)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

shouldn't this be doing this to the inner error and not the outer error?
</blockquote></details>

Yeah updated
___
*[`pkg/sql/synthesize_privileges.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_NzY1lTlbpuLaHnT:-NDirN-k2z6l0ldzhCR_:b-7uqh3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L101)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: consider pulling this for loop into a helper
</blockquote></details>

It doesn't make much sense to me to only pull the 
```
privileges = &catpb.PrivilegeDescriptor{}
			for {
```

into a helper.

Do you want me to pull the entire

```
func(ctx context.Context, ie sqlutil.InternalExecutor) (retErr error) {
				it, err := ie.QueryIteratorEx(ctx, `get-system-privileges`, txn,
					sessiondata.NodeUserSessionDataOverride, query)
				if err != nil {
					return err
				}
				defer func() {
					retErr = errors.CombineErrors(retErr, it.Close())
				}()

				privileges = &catpb.PrivilegeDescriptor{OwnerProto: username.AdminRoleName().EncodeProto()}
				for {
					ok, err := it.Next(ctx)
					if err != nil {
						return err
					}
					if !ok {
						break
					}
```
into a separate function?
___
*[`pkg/sql/catalog/descs/collection.go` line 141 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_hc486T0SyenGlSS:-NDis0-g8lkGb5-2cP_e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/catalog/descs/collection.go#L141)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

give this some commentary while you're here. What is this? What is it used for? etc. 
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/tabledesc/table_desc.go` line 650 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4aX3yDLPvEXKRCz5k:-NDiu0pWCR3euFR6-xXa:br420sj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/catalog/tabledesc/table_desc.go#L650)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you put commentary around this case? When does it happen? Should it happen? 
</blockquote></details>

Ah yeah this is for the mixed version case. If we're not on the `SystemPrivilegesTable` version, the desc will be nil, I suppose we could populate the descriptor at this point?

I'm gonna rmeove the logic from here and make it so we still fill out the priv desc in the same place we normally would with the "default" case.



_Code quote:_
```Go
username.PublicRoleName(), privilege.List{privilege.USAGE},
```
___
*[`pkg/sql/sem/eval/deps.go` line 345 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4aicA2Erwnqrqu_DG:-NDiowl2E5B11B4nLBCE:brmnndl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/sem/eval/deps.go#L345)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What does this get used for? 
</blockquote></details>

Don't need this, removed



_Code quote:_
```Go
SynthesizePrivilegeDescriptor
```
___
*[`pkg/sql/syntheticprivilege/privilege_synthesizer.go` line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4aq6AEKy350FrM3oe:-NDiozAyCrzfetwlWNw_:b-qat6t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/syntheticprivilege/privilege_synthesizer.go#L28)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

what is this about? 
</blockquote></details>

Did this to get around `server_sql.go`'s order of initializing things. `privilegeSynthesizer` is passed into `collectionFactory` which is needed for `execCfg` which used to create `pgServer` and `pgServer` is used to create the monitor needed for the `ieFactory`.

We could probably untangle this but we were doing this for `jobRegistry` anyway

```	
privilegeSynthesizer.SetIEFactory(ieFactory)

distSQLServer.ServerConfig.InternalExecutorFactory = ieFactory
jobRegistry.SetInternalExecutorFactory(ieFactory)
```
___
*[`pkg/ccl/backupccl/show.go` line 293 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4ZkyX40qultePy4UF:-NDimx8W9QnbXFFaXgNs:b-v2ewjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/ccl/backupccl/show.go#L293)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Isn't mutable here wrong? Should this be immutable? Mutable means avoiding leases. 
</blockquote></details>

Actually I believe the rest of the refactor (eagerly resolving the privilege desc) fixes at least the test case that I added and the `SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'] UNION SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'];` case 

Ran under stress and it does not fail.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88787)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin and @RichardJCai)

___
*[`pkg/ccl/backupccl/show.go` line 293 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4ZkyX40qultePy4UF:-NDiygdF523uc4y3DX85:b8qa3lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/ccl/backupccl/show.go#L293)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Actually I believe the rest of the refactor (eagerly resolving the privilege desc) fixes at least the test case that I added and the `SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'] UNION SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'];` case 

Ran under stress and it does not fail.
</blockquote></details>

By wrong I didn't mean buggy I meant inefficient. 


<!-- Sent from Reviewable.io -->

RichardJCai commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88787)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin and @RichardJCai)
> 
> _[`pkg/ccl/backupccl/show.go` line 293 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4ZkyX40qultePy4UF:-NDiygdF523uc4y3DX85:b8qa3lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/ccl/backupccl/show.go#L293)):_
> 
> Previously, RichardJCai (Richard Cai) wrote…
> > Actually I believe the rest of the refactor (eagerly resolving the privilege desc) fixes at least the test case that I added and the `SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'] UNION SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'];` case
> > Ran under stress and it does not fail.
> 
> By wrong I didn't mean buggy I meant inefficient.

Ah, I've removed it completely now though, the resolve desc, it looks like we don't actually need it?

ajwerner reviewed (commented):
> Reviewed 12 of 44 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88787)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin and @RichardJCai)

___
*[`pkg/sql/authorization.go` line 117 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-NE20Hzp4yEcLpigzrxg:-NE20Hzp4yEcLpigzrxh:biljns2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e69988e6c8ede8ca62c8e134f360e0e03a74518/pkg/sql/authorization.go#L117)):*
> ```Go
> 	HasRoleOption(ctx context.Context, roleOption roleoption.Option) (bool, error)
> 
> 	SynthesizePrivilegeDescriptor(
> ```

nit: add commentary
___
*[`pkg/sql/synthesize_privileges.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-ND4_NzY1lTlbpuLaHnT:-NE20tH2-sxG7CB1M9X0:b-pfm7zw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9ac6c5ce87afc4d23eefc8f3ce0ace4ea189892/pkg/sql/synthesize_privileges.go#L101)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

It doesn't make much sense to me to only pull the 
```
privileges = &catpb.PrivilegeDescriptor{}
			for {
```

into a helper.

Do you want me to pull the entire

```
func(ctx context.Context, ie sqlutil.InternalExecutor) (retErr error) {
				it, err := ie.QueryIteratorEx(ctx, `get-system-privileges`, txn,
					sessiondata.NodeUserSessionDataOverride, query)
				if err != nil {
					return err
				}
				defer func() {
					retErr = errors.CombineErrors(retErr, it.Close())
				}()

				privileges = &catpb.PrivilegeDescriptor{OwnerProto: username.AdminRoleName().EncodeProto()}
				for {
					ok, err := it.Next(ctx)
					if err != nil {
						return err
					}
					if !ok {
						break
					}
```
into a separate function?
</blockquote></details>

This is all just very indented to the point that I've lost track of the context in which this logic is run.
___
*[`pkg/sql/synthesize_privileges.go` line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-NE20izR2NlnxHC2T56J:-NE20izR2NlnxHC2T56K:bzc4i3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e69988e6c8ede8ca62c8e134f360e0e03a74518/pkg/sql/synthesize_privileges.go#L57)):*
> ```Go
> 
> // SynthesizePrivilegeDescriptor is part of the Planner interface.
> func (p *PrivilegeSynthesizer) SynthesizePrivilegeDescriptor(
> ```

something here smells wrong to me. How do we check if this transaction has modified any of these privileges without a descs.Collection? Is it part of the Planner interface?
___
*[`pkg/sql/synthesize_privileges.go` line 93 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-NE21E6YBj7gQNerTvu0:-NE21E6YBj7gQNerTvu1:b1nc3jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e69988e6c8ede8ca62c8e134f360e0e03a74518/pkg/sql/synthesize_privileges.go#L93)):*
> ```Go
> 
> 			err := p.ieFactory.RunWithoutTxn(loadCtx, func(ctx context.Context, ie sqlutil.InternalExecutor) (retErr error) {
> 				it, err := ie.QueryIteratorEx(ctx, `get-system-privileges`, txn,
> ```

are you allowed to pass a transaction to an ie created for `RunWithoutTxn`? @zhouxing19 what's the intended semantics here?
___
*[`pkg/sql/synthesize_privileges.go` line 195 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-NE21S70-B3WXzXZ8dB5:-NE21S70-B3WXzXZ8dB6:bi2yyou) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e69988e6c8ede8ca62c8e134f360e0e03a74518/pkg/sql/synthesize_privileges.go#L195)):*
> ```Go
> 		return nil, errors.AssertionFailedf("failed to find system.privileges table")
> 	}
> 	return p.privilegeSynthesizer.SynthesizePrivilegeDescriptor(ctx, p.Txn(), path, privilegeObjectType, desc.GetVersion())
> ```

don't you need to do something special if this transaction has modified the descriptor? It'd be good to see tests blow up in some way if this assumption is violated. 
___
*[`pkg/sql/catalog/descs/descriptor.go` line 420 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-NE21kuf1QvA1WYQ6RMx:-NE21kuf1QvA1WYQ6RMy:b-y4et4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e69988e6c8ede8ca62c8e134f360e0e03a74518/pkg/sql/catalog/descs/descriptor.go#L420)):*
> ```Go
> 				if err != nil {
> 					return haltLookups, nil, err
> 				}
> ```

can all of this logic be encapsulated? Doesn't it need to do somethign special if the privileges have been modified?



_Code quote:_
```Go
			if !tc.settings.Version.IsActive(ctx, clusterversion.SystemPrivilegesTable) {
				found, desc, err := tc.GetImmutableTableByName(
					ctx, txn, syntheticprivilege.SystemPrivilegesTableName, tree.ObjectLookupFlags{},
				)
				if err != nil {
					return haltLookups, nil, err
				}
				if !found {
					return haltLookups, nil, errors.AssertionFailedf("failed to find system.privileges table")
				}
				privDesc, err = tc.privilegeSynthesizer.SynthesizePrivilegeDescriptor(
					ctx, txn, syntheticprivilege.CreateVirtualTablePrivilegePath(sc.GetName(), vd.GetName()),
					privilege.VirtualTable, desc.GetVersion(),
				)
				if err != nil {
					return haltLookups, nil, err
				}
```


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'] UNION SELECT * FROM [SHOW BACKUPS IN 'nodelocal://0/test'];

Can you help me understand more concretely how we avoid resolving the synthetic privileges for the different urls concurrently? Would it matter if the union all was over different URLs?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88787#-:-NE227Gd04RhkzB3Qsuz:bqqkpvm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin and @RichardJCai)



<!-- Sent from Reviewable.io -->

RichardJCai committed:
sql: some more refactors

Release note: None

RichardJCai commented:
I'm fairly busy this week but I'm happy to give this some cleanup next weekend. 

Need to setup my Goland again :/

If y'all want to get this in sooner, feel free to take it over.

postamar commented:
No worries, I'll take it over.
